core/shared/src/main/scala/spire/algebra/Gcd.scala

```diff
@@ -0,0 +1,25 @@
+package spire.algebra
+
+import annotation.tailrec
+import scala.{specialized => sp}
+
+import spire.syntax.eq._
+import spire.syntax.euclideanRing._
+
+trait Gcd[@sp(Int, Long, Float, Double) A] extends Any {
+  def gcd(a: A, b: A): A
+  def lcm(a: A, b: A): A
+
+  @tailrec final def euclid(a: A, b: A)(implicit eq: Eq[A], er: EuclideanRing[A]): A =
+    if (b === er.zero) a else euclid(b, a % b)
+}
+
+object Gcd {
+  @inline final def apply[A](implicit ev: Gcd[A]): Gcd[A] = ev
+
+  implicit def fromEuclideanRing[@sp(Int, Long, Float, Double) A: EuclideanRing: Eq] =
```

Comment:
@tixxit If `EuclideanRing` and `Gcd` are defined properly, there will be no edge cases when dealing with polynomials.

I will have a look later today. The definition of `EuclideanRing` in `algebra` is fine, I will have a look at the proposed `Gcd` typeclass --- and all the default instances defined.

I will also make sure the documentation is clear, because:
- the mathematical definitions on Wikipedia (our default encyclopedia) are extremely rich (commutative rings ⊃ integral domains ⊃ integrally closed domains ⊃ GCD domains ⊃ unique factorization domains ⊃ principal ideal domains ⊃ Euclidean domains ⊃ fields ⊃ finite fields), not joking!
- the function `fmod(lhs, rhs)` does not correspond to a definition in the chain above (I interpret `fmod` as: return `lhs - k * rhs` where `k` is the maximal integer such that the result is positive). Still, we call often it "modulo", and do not see the difference with the modulo operation that appears in Euclid's algorithm.